trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/collections/MultiMap.java

```diff
@@ -0,0 +1,226 @@
+        }
+        ensureKeyPresent(key).addAll(values);
+    }
+
+    /**
+     * Obtain a copy of the keys of the data in the map
+     *
+     * @return Immutable copy of the keys of the map
+     */
+    public Set<String> keySet()
+    {
+        return ImmutableSet.copyOf(mapData.keySet());
+    }
+
+    /**
+     * Obtain the set of entries in the map
+     */
+    public Set<Map.Entry<String, List<V>>> entrySet()
+    {
+        return mapData.entrySet();
+    }
+
+    /**
+     * Get all values associated to a given key in the map.
+     * Returns null if the key is not present in the map.
+     *
+     * @param key key
+     * @return list of values, null if key not present
+     * @throws NullPointerException if key is null
+     */
+    public List<V> get(String key)
+    {
+        return mapData.get(getActualKey(key));
+    }
+
+    /**
+     * Obtain the first value associated with a key, null if no values are associated
+     *
+     * @return first value associated with the key, null if none
+     * @throws NullPointerException if key is null
+     */
+    public V getFirst(String key)
+    {
+        if (!containsKey(key)) {
+            return null;
+        }
+        List<V> values = get(key);
+        return values.isEmpty() ? null : values.getFirst();
+    }
+
+    /**
+     * Process each entry in the map using the provided consumer
+     *
+     * @param consumer function to be executed for each entry
+     */
+    public void forEach(BiConsumer<String, List<V>> consumer)
+    {
+        mapData.forEach(consumer);
+    }
+
+    /**
+     * Get the list of values associated with a key, if any. Otherwise, return a default
+     *
+     * @param key key to retrieve
+     * @param defaultValue default value to return if no values are associated with the provided key
+     * @return List of values associated with the key (if any), default value otherwise
+     * @throws NullPointerException if key is null
+     */
+    public List<V> getOrDefault(String key, Collection<V> defaultValue)
+    {
+        if (!containsKey(key)) {
+            return newCollectionInstance(defaultValue);
+        }
+        return mapData.get(getActualKey(key));
+    }
+
+    /**
+     * Determine whether a given key as at least one value associated with it in the map
+     *
+     * @throws NullPointerException if key is null
+     */
+    public boolean containsKey(String key)
+    {
+        String actualKey = getActualKey(key);
+        return mapData.containsKey(actualKey) && !mapData.get(actualKey).isEmpty();
+    }
+
+    private Collection<V> ensureKeyPresent(String key)
+    {
+        String actualKey = getActualKey(key);
+        if (!mapData.containsKey(actualKey)) {
+            mapData.put(actualKey, newCollectionInstance());
+        }
+        return mapData.get(actualKey);
+    }
+
+    public static <V> MultiMap<V> copyOf(MultiMap<V> values)
+    {
+        return copyOf(values.entrySet(), values.isCaseSensitiveKeys());
```

Comment:
Check if `values` is an instance of `ImmutableMultiMap`. If so, you can just return the argument unchanged for a no-cost "copy".